src/Dictionaries/LibraryDictionarySource.cpp

```diff
@@ -8,292 +9,182 @@
-            auto data_ref = key_columns[i]->getDataAt(requested_rows[j]);
-            cell_holder[j] = ClickHouseLibrary::Field{.data = static_cast<const void *>(data_ref.data), .size = data_ref.size};
-        }
-        holder[i]
-            = ClickHouseLibrary::Row{.data = static_cast<ClickHouseLibrary::Field *>(cell_holder.get()), .size = requested_rows.size()};
-
-        column_data_holders.push_back(std::move(cell_holder));
-    }
+        std::string key_name = key;
+        auto bracket_pos = key.find('[');
 
-    ClickHouseLibrary::Table request_cols{.data = static_cast<ClickHouseLibrary::Row *>(holder.get()), .size = key_columns.size()};
+        if (bracket_pos != std::string::npos && bracket_pos > 0)
+            key_name = key.substr(0, bracket_pos);
 
-    auto load_keys_func = library->get<ClickHouseLibrary::LibraryLoadKeysFunc>(ClickHouseLibrary::LIBRARY_LOAD_KEYS_FUNC_NAME);
-    auto data_new_func = library->get<ClickHouseLibrary::LibraryDataNewFunc>(ClickHouseLibrary::LIBRARY_DATA_NEW_FUNC_NAME);
-    auto data_delete_func = library->get<ClickHouseLibrary::LibraryDataDeleteFunc>(ClickHouseLibrary::LIBRARY_DATA_DELETE_FUNC_NAME);
+        if (res.stringRef().size)
+            writeChar(' ', res);
 
-    ClickHouseLibrary::LibraryData data_ptr = data_new_func(lib_data);
-    SCOPE_EXIT(data_delete_func(lib_data, data_ptr));
-
-    ClickHouseLibrary::RawClickHouseLibraryTable data = load_keys_func(data_ptr, &settings->strings, &request_cols);
-    auto block = dataToBlock(description.sample_block, data);
+        writeString(key_name, res);
+        writeChar(' ', res);
+        writeString(config.getString(config_root + "." + key), res);
+    }
 
-    return std::make_shared<OneBlockInputStream>(block);
+    return res.str();
 }
 
-bool LibraryDictionarySource::isModified() const
-{
-    if (auto func_is_modified = library->tryGet<ClickHouseLibrary::LibraryIsModifiedFunc>(
-        ClickHouseLibrary::LIBRARY_IS_MODIFIED_FUNC_NAME))
-        return func_is_modified(lib_data, &settings->strings);
-
-    return true;
-}
 
-bool LibraryDictionarySource::supportsSelectiveLoad() const
+String LibraryDictionarySource::getDictAttributesString()
 {
-    if (auto func_supports_selective_load = library->tryGet<ClickHouseLibrary::LibrarySupportsSelectiveLoadFunc>(
-            ClickHouseLibrary::LIBRARY_SUPPORTS_SELECTIVE_LOAD_FUNC_NAME))
-        return func_supports_selective_load(lib_data, &settings->strings);
+    WriteBufferFromOwnString res;
+    for (const auto & attr : dict_struct.attributes)
+    {
+        if (res.stringRef().size)
+            writeChar(' ', res);
+        writeString(attr.name, res);
+    }
 
-    return true;
+    return res.str();
 }
 
-DictionarySourcePtr LibraryDictionarySource::clone() const
-{
-    return std::make_unique<LibraryDictionarySource>(*this);
-}
 
-std::string LibraryDictionarySource::toString() const
+String LibraryDictionarySource::getDictIdsString(const std::vector<UInt64> & ids)
```

Comment:
I think we should avoid some custom space-separated format and just serialize them to POST body in some already implemented format.